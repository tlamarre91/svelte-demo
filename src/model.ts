import { generateSlug } from "random-word-slugs";

export class Participant {
  name: string;
  createdAt: Date;
  lastModifiedAt: Date;
  removedAt: Date | null;
  constructor(opts: Partial<Participant>) {
    // TODO: default name should probably be null.
    this.name = opts.name ?? "";
    this.createdAt = opts.createdAt ?? new Date();
    this.lastModifiedAt = opts.lastModifiedAt ?? new Date();
    this.removedAt = opts.removedAt ?? null;
  }

  static makeRandom() {
    return new Participant({
      name: generateSlug(2, { format: "title" }),
    });
  }
}

/**
 * A pending or completed match in a Bracket. A null participant field means
 * that a previous match hasn't been reported to determine the winner that will
 * proceed to this match.
 */
export class Match {
  /**
   * First participant. If pending is false, this is the winner.
   */
  participant1: number | null;
  /**
   * Second participant. If pending is false, this is the loser.
   */
  participant2: number | null;
  /**
   * If true, this is a yet-to-be-played match; otherwise, it is a match with a
   * winner and a loser.
   */
  pending: boolean;
  constructor(
    participant1?: number | null,
    participant2?: number | null,
    pending = true
  ) {
    if (!pending && !(participant1 != null && participant2 != null)) {
      console.log("values:", participant1, participant2, pending);
      throw new Error(
        "Match without both participants must have pending set to true."
      );
    }
    this.participant1 = participant1 ?? null;
    this.participant2 = participant2 ?? null;
    this.pending = pending;
  }
}

export class Bracket {
  name: string;
  slug?: string; // TODO: better handling of where slug gets generated
  userId: string;
  createdAt: Date;
  lastModifiedAt: Date;
  finalizedAt: Date | null;
  participants: Participant[];
  /**
   * Map from round number to a list of Matches in that round.
   */
  matches: Map<number, Match[]>;
  constructor(opts: Partial<Bracket>) {
    if (opts.userId == undefined) {
      // TODO: just define a type with some optional, some required args
      throw new Error("Bracket must have userId");
    }
    this.name = opts.name ?? "";
    this.slug = opts.slug;
    this.userId = opts.userId;
    this.finalizedAt = opts.finalizedAt ?? null;
    this.createdAt = opts.createdAt ?? new Date();
    this.lastModifiedAt = opts.lastModifiedAt ?? new Date();
    this.participants = opts.participants ?? [];
    this.matches = opts.matches ?? new Map();
  }

  /**
   * If we start the tournament with the current number of participants, how
   * many rounds will it last? Integer ceiling of log_2(this.participants.length)
   */
  computeTotalRounds() {
    const np = this.participants.length;
    if (np < 2) {
      return 0;
    } else {
      return Math.ceil(Math.log2(np));
    }
  }

  /**
   * If a match between these two participant indices, in this round, has been
   * recorded, return its index in the matches array. If not, return -1.
   */
  findMatch(p1: number, p2: number, round: number) {
    const roundMatches = this.matches.get(round);
    if (roundMatches == undefined) {
      return -1;
    }
    return roundMatches.findIndex(
      ({ participant1: winner, participant2: loser }: Match) => {
        return (p1 == winner && p2 == loser) || (p2 == winner && p1 == loser);
      }
    );
  }

  /**
   * In which (0-indexed!) round would a hypothetical match between these two participant
   * indices occur? Returns 0 if arguments are equal.
   *
   * This funky function works backwards from the root of the bracket (viewed
   * like a binary tree) and finds the first node where p1 is on one side and
   * p2 is on the other. I'm pretty sure there's a simpler way to do this.
   */
  computeMatchRound(p1: number, p2: number) {
    if (
      p1 < 0 ||
      p2 < 0 ||
      p1 >= this.participants.length ||
      p2 >= this.participants.length
    ) {
      throw new Error(`Illegal arguments: ${p1}, ${p2}`);
    }
    if (p1 == p2) {
      return 0;
    }

    const r = this.computeTotalRounds() - 1;
    for (let i = r; i >= 0; i -= 1) {
      const pow = Math.pow(2, i);
      const sideOfP1 = Math.floor(p1 / pow);
      const sideOfP2 = Math.floor(p2 / pow);
      if (sideOfP1 != sideOfP2) {
        return i;
      }
    }
    return 0;
  }

  /**
   * Does our Map of matches have a match for this participant in this round?
   *
   * @param round
   *   Key of this.matches to check.
   * @param participantIndex
   *   Index [0, ..., this.participants.length - 1] of participant in
   *   this.participants.
   * @param includePending
   *   If true, return true if match is pending. Otherwise, pending matches
   *   don't count.
   */
  matchExists(
    round: number,
    participantIndex: number,
    includePending: boolean = false
  ) {
    const roundMatches = this.matches.get(round);
    if (roundMatches == undefined) {
      return false;
    }
    return roundMatches.some(
      ({ participant1: winner, participant2: loser, pending }: Match) => {
        return (
          (includePending || !pending) &&
          (participantIndex == winner || participantIndex == loser)
        );
      }
    );
  }

  /**
   * Does our Map of matches have a *win* for this participant in this round?
   */
  winExists(round: number, participantIndex: number) {
    const roundMatches = this.matches.get(round);
    if (roundMatches == undefined) {
      return false;
    }
    return roundMatches.some(({ participant1: winner, pending }: Match) => {
      return !pending && participantIndex == winner;
    });
  }

  /**
   * Given the current Map of matches, have both of this match's participants
   * won the necessary rounds to meet each other? If either participant is null,
   * we say that the match is valid.
   */
  matchIsValid({ participant1: winner, participant2: loser }: Match) {
    if (winner == null || loser == null) return true;
    if (winner < 0 || loser < 0 || winner == loser) return false;

    const computedRound = this.computeMatchRound(winner, loser);
    // Check that we don't already have a match (pending or not) for either
    // participant in the appropriate round.
    const cond =
      this.matchExists(computedRound, winner) ||
      this.matchExists(computedRound, loser);
    if (cond) {
      return false;
    }
    // Check that both participants won matches in all the previous rounds.
    for (let round = 0; round < computedRound; round += 1) {
      const cond = !(
        this.winExists(round, winner) && this.winExists(round, loser)
      );
      if (cond) {
        return false;
      }
    }

    return true;
  }

  /**
   * Add a (pending) match to our Map of matches. Don't call this with matches
   * that have already been played. First, add a pending match, then call
   * reportMatch.
   */
  pushPendingMatch(match: Match, round: number) {
    if (!this.finalizedAt) {
      throw new Error(
        "Can't push pending match to bracket that hasn't been finalized."
      );
    }
    // const { participant1: p1, participant2: p2, pending } = match;
    if (!pending) {
      throw new Error("Can't add a non-pending match with pushPendingMatch.");
    }
    // const round = this.computeMatchRound(p1, p2);
    if (!this.matchIsValid(match)) {
      // TODO: matchIsValid should really tell us why a match isn't valid.
      throw new Error(`Can't push invalid match: ${match}`);
    }
    const roundMatches = this.matches.get(round);
    if (roundMatches == undefined) {
      this.matches.set(round, [match]);
    } else {
      roundMatches.push(match);
    }
  }

  /**
   * Update a pending match with the actual results of the match, after
   * checking that this operation is valid.
   *
   * @param match
   *   The match to report. Note: `pending` will be set to false, regardless of
   *   value of match[2].
   */
  reportMatch(match: Match) {
    if (!this.finalizedAt) {
      throw new Error(
        "Can't push match result to bracket that hasn't been finalized."
      );
    }

    const { participant1: p1, participant2: p2 } = match;
    if (p1 == null || p2 == null) {
      throw new Error(
        "Reporting a match requires both participants to be defined."
      );
    }
    const computedRound = this.computeMatchRound(p1, p2);
    const existingMatchIndex = this.findMatch(p1, p2, computedRound);

    if (existingMatchIndex == -1) {
      const err = `Pending match not found in round ${computedRound} for participants ${p1} and ${p2}`;
      throw new Error(err);
    }

    const roundMatches = this.matches.get(computedRound);
    if (roundMatches == undefined) {
      // TODO: this check is probably not necessary, since we just called
      // findMatch and it succeeded.
      throw new Error(`Error finding match ${match} in round ${computedRound}`);
    }

    const existingMatch = roundMatches[existingMatchIndex];
    const { pending: ePending } = existingMatch;
    if (!ePending) {
      throw new Error(
        `Can't report match: match ${existingMatch} in round ${computedRound} is already reported.`
      );
    }

    Object.assign(existingMatch, { ...match, pending: false });
  }

  /**
   * Validate that this bracket is ready to be finalized, meaning that we're
   * ready to start pushing match results.
   *
   * @param validateOnly
   *   Just check that this bracket is ready to be finalized. Don't actually
   *   set finalizedAt.
   */
  finalize(validateOnly = false) {
    if (this.finalizedAt) {
      return; // TODO: should this throw?
    }

    const nParticipants = this.participants.length;
    const lp = Math.log2(nParticipants);
    if (!Number.isInteger(lp)) {
      const nextPow = Math.pow(2, Math.ceil(lp));
      const err = `Number of participants must be a power of 2. (Current count: ${nParticipants}. Add ${
        nextPow - nParticipants
      } more.)`;
      throw new Error(err);
    }

    if (validateOnly) {
      return;
    }

    this.finalizedAt = new Date();
  }

  /**
   * Unset finalizedAt and clear all matches.
   */
  unfinalize() {
    if (!this.finalizedAt) {
      throw new Error("Can't unfinalize a bracket that isn't finalized.");
    }
    this.finalizedAt = null;
    this.matches = new Map();
  }
}
