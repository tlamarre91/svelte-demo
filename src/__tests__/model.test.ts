import * as model from "../model";
import { pruneDates } from "./helpers";

const names = [
  "Skinny Balloon",
  "Odd Library",
  "Savory Candle",
  "Square Rose",
  "Breezy Truck",
  "Wrong Restaurant",
  "Moldy Guitar",
  "Cuddly Planet",
];

function makeParticipantsFromNames() {
  return names.map((name) => new model.Participant({ name }));
}

describe("Bracket model", () => {
  let bracket: model.Bracket;
  const userId = "testId";
  const name = "testName";

  beforeEach(() => {
    const participants = makeParticipantsFromNames();
    bracket = new model.Bracket({ userId, name, participants });
  });

  test("Smoke test", () => {
    expect(bracket.userId).toBe(userId);
    expect(bracket.name).toBe(name);
    expect(bracket.participants.map(pruneDates)).toMatchSnapshot();
  });

  test("findMatch", () => {
    const matches: model.Match[] = [
      new model.Match(0, 1, true),
      new model.Match(3, 2, true),
      new model.Match(0, 3, true),
    ];
    bracket.finalize();
    bracket.pushPendingMatch(matches[0], 0);
    bracket.reportMatch(matches[0]);
    bracket.pushPendingMatch(matches[1], 0);
    bracket.reportMatch(matches[1]);
    bracket.pushPendingMatch(matches[2], 1);
    expect(bracket.findMatch(0, 1, 0)).toBe(0);
    expect(bracket.findMatch(3, 2, 0)).toBe(1);
    expect(bracket.findMatch(0, 3, 1)).toBe(0);
    expect(bracket.findMatch(1, 3, 0)).toBe(-1);
  });

  test("computeTotalRounds", () => {
    expect(bracket.computeTotalRounds()).toBe(3);
    bracket.participants.push(...makeParticipantsFromNames());
    // Double the participant count => one more round.
    expect(bracket.computeTotalRounds()).toBe(4);
    bracket.participants.push(...makeParticipantsFromNames());
    bracket.participants.push(...makeParticipantsFromNames());
    expect(bracket.computeTotalRounds()).toBe(5);
  });

  test("computeMatchRound throws with out-of-bounds arguments", () => {
    expect(() =>
      bracket.computeMatchRound(-1, 2)
    ).toThrowErrorMatchingInlineSnapshot(`"Illegal arguments: -1, 2"`);
    expect(() =>
      bracket.computeMatchRound(3, -1)
    ).toThrowErrorMatchingInlineSnapshot(`"Illegal arguments: 3, -1"`);
    expect(() =>
      bracket.computeMatchRound(3, 8)
    ).toThrowErrorMatchingInlineSnapshot(`"Illegal arguments: 3, 8"`);
    expect(() =>
      bracket.computeMatchRound(8, 3)
    ).toThrowErrorMatchingInlineSnapshot(`"Illegal arguments: 8, 3"`);
  });

  test("computeMatchRound with 8 participants", () => {
    // Worked out a few test cases by hand:
    expect(bracket.computeMatchRound(0, 0)).toBe(0);
    expect(bracket.computeMatchRound(0, 1)).toBe(0);
    expect(bracket.computeMatchRound(0, 2)).toBe(1);
    expect(bracket.computeMatchRound(2, 5)).toBe(2);
    expect(bracket.computeMatchRound(0, 7)).toBe(2);
  });

  test("computeMatchRound with 16 participants", () => {
    bracket.participants.push(...bracket.participants);
    // These cases are the same as when participants.length == 8:
    expect(bracket.computeMatchRound(0, 0)).toBe(0);
    expect(bracket.computeMatchRound(0, 1)).toBe(0);
    expect(bracket.computeMatchRound(0, 2)).toBe(1);
    expect(bracket.computeMatchRound(2, 5)).toBe(2);
    expect(bracket.computeMatchRound(0, 7)).toBe(2);

    // Check some new cases:
    expect(bracket.computeMatchRound(7, 8)).toBe(3);
    expect(bracket.computeMatchRound(8, 9)).toBe(0);
    expect(bracket.computeMatchRound(11, 12)).toBe(2);
    expect(bracket.computeMatchRound(14, 15)).toBe(0);
    expect(bracket.computeMatchRound(4, 10)).toBe(3);
  });

  test("winExists returns true when participant has a win in the specified round", () => {
    const p1 = 0;
    const p2 = 1;
    bracket.finalize();
    bracket.pushPendingMatch(new model.Match(p1, p2, true), 0);
    bracket.reportMatch(new model.Match(p1, p2, false));
    expect(bracket.winExists(0, p1)).toBe(true);
  });

  test("winExists returns false when participant does not have a win in the specified round", () => {
    const p1 = 0;
    const p2 = 1;
    bracket.finalize();
    bracket.pushPendingMatch(new model.Match(p1, p2, true), 0);
    bracket.reportMatch(new model.Match(p1, p2, false));
    expect(bracket.winExists(0, p2)).toBe(false);
  });

  test("matchIsValid correctly identifies when a match is valid", () => {
    bracket.finalize();
    const m1: model.Match = new model.Match(0, 2, true);
    // These participants don't meet before they've played 1 round each.
    expect(bracket.matchIsValid(m1)).toBe(false);
    bracket.pushPendingMatch(new model.Match(0, 1, true), 0);
    bracket.pushPendingMatch(new model.Match(2, 3, true), 0);
    bracket.reportMatch(new model.Match(0, 1, false));
    bracket.reportMatch(new model.Match(2, 3, false));
    expect(bracket.matchIsValid(m1)).toBe(true);
    bracket.pushPendingMatch(m1, 1);
    // bracket.reportMatch([m1[0], m1[1], true]);
    bracket.reportMatch(
      new model.Match(m1.participant1, m1.participant2, false)
    );
    // These participants already played their match; the same match won't be valid again.
    expect(bracket.matchIsValid(m1)).toBe(false);
  });

  test("finalize throws when participant count is not power of 2", () => {
    bracket.participants.push(new model.Participant({ name: "New guy" }));
    expect(() => bracket.finalize()).toThrowErrorMatchingInlineSnapshot(
      `"Number of participants must be a power of 2. (Current count: 9. Add 7 more.)"`
    );
  });

  test("finalize succeeds with a well-formed bracket", () => {
    bracket.finalize();
    const diff = Math.abs(Date.now() - bracket.finalizedAt?.getTime()!);
    expect(diff).toBeLessThanOrEqual(50);
  });

  test("finalize doesn't actually do anything when validateOnly is true", () => {
    bracket.finalize(true);
    expect(bracket.finalizedAt).toBeNull();
  });
});
