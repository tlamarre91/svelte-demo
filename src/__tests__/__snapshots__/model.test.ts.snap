// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bracket model Smoke test 1`] = `
Array [
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Skinny Balloon",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Odd Library",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Savory Candle",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Square Rose",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Breezy Truck",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Wrong Restaurant",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Moldy Guitar",
    "removedAt": null,
  },
  Object {
    "createdAt": "[date]",
    "lastModifiedAt": "[date]",
    "name": "Cuddly Planet",
    "removedAt": null,
  },
]
`;
